In [88]:
print("Hello! Python!")

Hello! Python!


In [89]:
class Person:
    def __init__(self,name):#自訂的初始化
        self.name = name  #attribute
    
    def greeting(self):#自訂的實體方法
        print(f"您好{self.name}")
    



In [90]:
p1: Person = Person("robert")
p1.greeting()   #  執行方法 → 您好robert

p2: Person = Person(name="jenny")
p2.greeting()   #  執行方法 → 您好jenny

您好robert
您好jenny


In [91]:
# 定義父類別 Person
class Person:
    def __init__(self, name):  # 初始化方法，建立物件時會自動執行
        self.name = name       # 將名字存到屬性 name
    
    def greeting(self):        # 一般方法，印出問候語
        print(f"您好 {self.name}")


# 定義子類別 student，繼承自 Person
class student(Person):
    def __init__(self, name, chinese, english, math):
        # 呼叫父類別(Person)的初始化，先把 name 存起來
        super().__init__(name)
        # 再加上三個屬性：國文、英文、數學 (前面加 __ 代表私有變數)
        self.__chinese = chinese
        self.__english = english
        self.__math = math
    
    # getter (用 @property 包裝，讓呼叫端像讀屬性一樣取值)
    @property
    def chinese(self):
        return self.__chinese
    
    @property
    def english(self):
        return self.__english

    @property
    def math(self):
        return self.__math

    # 覆寫父類別的 greeting()
    def greeting(self):
        super().greeting()   # 先呼叫父類別的 greeting()
        print("您好:我是學生")  # 再加上學生自己的問候
    
    # 計算總分的方法
    def total(self):
        return self.chinese + self.english + self.math
    
    # 計算平均的方法
    def average(self):
        return self.total() / 3.0
    
    # 當 print(s1) 時，自動呼叫這裡
    def __repr__(self):
        return f"""
        我是一個學生
        我的名字是{self.name}
        我的總分是{self.total()}
        我的平均是{self.average()}
        """

In [92]:
# 建立一個 student 物件，名字 kevin，三科分數：國文100、英文78、數學92
s1 = student("kevin", english=78, math=92, chinese=100)
# 印出名字 (來自父類別)
print(s1.name)     
# 取用國英數三科分數 (因為有用 @property，所以看起來像讀屬性)
print(f"國文分數:{s1.chinese}")
print(f"英文分數:{s1.english}")
print(f"數學分數:{s1.math}")

# 呼叫方法算總分和平均
print(f"總分:{s1.total()}")
print(f"平均:{s1.average()}")

# 呼叫 greeting()，會先執行父類別的 → "您好 kevin"
# 再執行子類別的 → "您好:我是學生"
s1.greeting()  

kevin
國文分數:100
英文分數:78
數學分數:92
總分:270
平均:90.0
您好 kevin
您好:我是學生


In [93]:
print(s1)


        我是一個學生
        我的名字是kevin
        我的總分是270
        我的平均是90.0
        


📝 總結：這段程式在做什麼？

這是一個示範 類別 (class) 與物件 (object) 的程式。
它分成 父類別 Person 和 子類別 student，子類別繼承父類別，並且加上自己的功能。

1. Person 類別（父類別）

負責存放名字 (name)。

有一個 greeting() 方法，會印出「您好 + 名字」。

👉 可以想像 Person 就是「一個會打招呼的人」。

2. student 類別（子類別，繼承 Person）

除了會打招呼，還會多存三個成績：國文、英文、數學。

@property 讓你可以直接用 s1.chinese 這樣的方式去讀分數（不用再寫方法呼叫）。

total()：算總分。

average()：算平均。

greeting()：改寫（覆寫）父類別的方法。
→ 先執行父類別的「您好 kevin」，再加上自己的「您好:我是學生」。

__repr__()：當你 print(s1) 的時候，會輸出「我是學生 + 名字 + 總分 + 平均」。

👉 可以想像 student 就是「一個學生，既能打招呼，也有考試分數」。

3. 程式執行流程
s1 = student("kevin", english=78, math=92, chinese=100)


建立一個學生物件 kevin，他的分數是：國文 100、英文 78、數學 92。

接著程式輸出：

印名字 → kevin

印三科成績 → 100, 78, 92

印總分 → 270

印平均 → 90.0

呼叫 greeting() →

「您好 kevin」 (來自 Person)

「您好:我是學生」 (來自 student)

如果加上 print(s1)，還會輸出：

我是學生
我的名字是 kevin
我的總分是 270
我的平均是 90.0

✅ 超新手用一句話理解

這個程式就是在示範：

Person = 一般人，只會打招呼。

student = 學生，不只會打招呼，還能存放分數、算總分與平均。

print(s1) 甚至會直接幫你顯示完整的學生資訊。